File: change_stream/change_stream.js
/**
 * change_stream.js
 * Author: Ken Chen
 */
var ascoltatori = require('ascoltatori');
const MongoClient = require("mongodb").MongoClient;
const assert = require("assert");

const pipeline = [
];

settings = {
    type: 'mqtt',
    json: false,
    mqtt: require('mqtt'),
    url: 'mqtt://test.mosquitto.org:1883'
};

function initMQTT() {
    return new Promise(function(resolve, reject) {
        ascoltatori.build(settings, function (err, ascoltatore) {
            if(err) {
                reject(err);
            } else {
                resolve(ascoltatore);
            }
        });
    });
}

initMQTT().then(mq => {
    MongoClient.connect("mongodb://localhost:27017/test?replicaSet=replset")
        .then(client => {
            console.log("Connected to server");
            const db = client.db("orders");
            const collection = db.collection("products");
            const changeStream = collection.watch(pipeline);
    
            changeStream.on("change", function(change) {
                payload = JSON.stringify({_id: change.documentKey._id, op: change.operationType, lang: "node.js"});
                console.log(payload);
                mq.publish('simagix/cs/test', payload, function() {
            });
        }); })
        .catch(err => {
        console.error(err);
    });
});

File: change_stream/change_stream.py
#! /usr/bin/env python
# change_stream.py
# Author: Ken Chen
#   3/31/2018 - first created
#

import json
import paho.mqtt.publish as publish
import pymongo
from pymongo import MongoClient
import sys
import time

client = None
db = None
collection = None

def init(mongo_uri):
    global client
    global db
    global collection
    client = MongoClient(mongo_uri)
    db = client.orders
    collection = db.products

if __name__ == "__main__":
    mongo_uri = 'mongodb://localhost:27017/test?replicaSet=replset'
    if len(sys.argv) > 1:
        mongo_uri = sys.argv[1]
    init(mongo_uri)
    try:
        for change in collection.watch():
            doc = {"op": change['operationType'], "_id":  change['documentKey']['_id'], "lang": "python"}
            payload = json.dumps(doc)
            print(payload)
            publish.single("simagix/cs/test", payload, hostname="test.mosquitto.org")
    except pymongo.errors.PyMongoError:
        console.log("error")
        pass

